core/classifiers/CodeClassifier/CodeClassifier.py

```diff
@@ -0,0 +1,710 @@
+                    'k_gram_hashes': list(int). K-gram hash values of program.
+                    'fingerprint': list(int). Extracted fingerprint of program.
+                    'top_similar': list(int). Most similar programs to this
+                        program.
+                } for each program in training_data.
+            }
+    """
+    for program_id_1 in training_data:
+        overlaps = []
+        for program_id_2 in training_data:
+            overlap = get_program_similarity(
+                training_data[program_id_1]['fingerprint'],
+                training_data[program_id_2]['fingerprint'])
+            overlaps.append((program_id_2, overlap))
+        training_data[program_id_1]['top_similar'] = sorted(
+            overlaps, key=lambda e: e[1], reverse=True)[:top]
+    return training_data
+
+
+def run_knn(training_data, top_neighbours):
+    """Predict classes for each program in dataset using KNN.
+
+    Args:training_data
+        training_data: dict. A dictionary containing training data.
+            Structure of training_data is as follows,
+                {
+                    ID: {
+                        'source': str. Source code of program.
+                        'class': int. Associated feedback class with program.
+                        'tokens': list(str). Token list of program.
+                        'k_gram_hashes': list(int). K-gram hash values of
+                            program.
+                        'fingerprint': list(int). Extracted fingerprint of
+                            program.
+                        'top_similar': list(int). Most similar programs to this
+                            program.
+                    } for each program in training_data.
+                }
+        top_neighbours: int. No. of nearest neighbours to consider for KNN.
+
+    Returns:
+        float. Average no. of times a class has to appear in nearest neighbour
+            for it to be correct prediction.
+        list(int). IDs of training_data points which are misclassified by KNN.
+    """
+    training_data = add_top_similars(training_data, top_neighbours)
+
+    # No. of times a class has to appear in nearest neighbours of a program
+    # so that prediction is correct.
+    occurrence = 0
+
+    # Keep record of misclassified programs' ID.
+    misclassified_points = []
+
+    for pid in training_data:
+        similars = training_data[pid]['top_similar']
+        if pid in similars:
+            similars.remove(pid)
+        nearest_classes = [training_data[i]['class'] for (i, _) in similars]
+        cnt = collections.Counter(nearest_classes)
+        common = cnt.most_common(1)
+        training_data[pid]['prediction'] = common[0][0]
+
+        if training_data[pid]['prediction'] == training_data[pid]['class']:
+            # If prediction is correct then add no. of times the correct
+            # class has to appear in nearest neighbours to occurrence.
```

Comment:
I guess I don't understand the "has to" part, then. Why do you say "the number of times the correct class has to appear in nearest neighbours" as opposed to "the number of times the correct class appears in the program's nearest neighbours"?